File: GW3/LogFileHandling/Query/Tokens/TokenDescending.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GWLogger.Backend.DataContext.Query.Tokens
{
    internal class TokenDescending : Token
    {
        public override bool CanBeUsed(QueryParser parser)
        {
            parser.SkipSpaces();

            return parser.PeekString().ToLower() == "desc";
        }

        public override Token Extract(QueryParser parser)
        {
            parser.SkipSpaces();
            parser.NextString();
            return new TokenDescending();
        }
    }
}

File: GW3/GWLogger/Frontend/Query/Tokens/TokenDescending.ts
class TokenDescending extends Token {

    public tokenType: TokenType = TokenType.TokenDescending;

    public proposals: SuggestionInterface[] = <SuggestionInterface[]>[{ suggestion: "desc" }];

    public canBeUsed(parser: QueryParser): boolean {
        parser.skipSpaces();
        return parser.peekString().toLowerCase() == "desc";
    }

    public extract(parser: QueryParser): Token {
        parser.skipSpaces();
        parser.nextString();
        let token = new TokenDescending();
        token.value = "desc";
        return token;
    }

    public getProposals(nextToken?: Token, afterNextToken?: Token): Token[] {
        if (typeof nextToken != 'undefined' && typeof afterNextToken != 'undefined') {
            switch (nextToken.tokenType) {
                case TokenType.TokenGroup:
                case TokenType.TokenOrder:
                    return [new TokenName()];
                case TokenType.TokenLimit:
                    return [new TokenNumber()];
                case TokenType.TokenWhere:
                    return [new TokenName(), new TokenOpenParenthesis()];
            }
        }
        return [];
    }
}